src/Microsoft.AspNetCore.Mvc.Core/Routing/ActionConstraintMatcherPolicy.cs

```diff
@@ -171,8 +205,10 @@ public Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context,
             var endpointsWithConstraint = new List<(int index, ActionSelectorCandidate candidate)>();
             var endpointsWithoutConstraint = new List<(int index, ActionSelectorCandidate candidate)>();
 
-            var constraintContext = new ActionConstraintContext();
-            constraintContext.Candidates = items.Select(i => i.candidate).ToArray();
+            var constraintContext = new ActionConstraintContext
+            {
+                Candidates = items.Select(i => i.candidate).ToArray()
```

Comment:
I know that this code is called as a fallback in cases when functionality trumps perf, but there are a bunch of per-request allocations here that are easy to improve 😬

Reply:
The contract here requires a bunch of allocations. That's why the routing one is so different. I'm not super passionate about improving the perf because we don't have a meaningful test for it. There are lots of areas of MVC where do care and they are useful scenarios 😆